include/tiramisu/core.h

```diff
@@ -136,6 +136,27 @@ void init();
 void codegen(const std::vector<tiramisu::buffer *> &arguments, const std::string obj_filename, const bool gen_cuda_stmt = false);
 void codegen(const std::vector<tiramisu::buffer *> &arguments, const std::string obj_filename, const tiramisu::hardware_architecture_t gen_architecture_flag);
 
+/**
+ * Full check of legality for this function using dependency analysis 
+ * must be used after invoking : performe_full_dependecy_analysis()
+ */
+bool check_legality_of_function();
+
+
+/**
+ * Performe a full dependecy analysis RAW/WAR/WAW and the result is stored in attributes inside the function
+ * To invoke this method user must : define computations order & define the buffer mapped to each computation using :
+ *   tiramisu::prepare_schedules_for_legality_checks()
```

Comment:
calling tiramisu::prepare_schedules_for_legality_checks() before each legality check (bool check_legality_of_function()) would reduce complexity. there is 2 motivations behind extraction out prepare_schedules_for_legality_checks() :

1. heavy calculation : since prepare_schedules_for_legality_checks() would be required to go through all the computations, and for each one apply a range of make them of the same out dimension. in the regular case this is okey, but what would happen if creating my space would require millions of checks with a high number of computation.
2. small merit : prepare_schedules_for_legality_checks() is only necessary when we apply an optimization that changes the out dimension of a schedule (tiling) or we fuse a computation in a different loop level.

My solution was to let the user manually call tiramisu::prepare_schedules_for_legality_checks() .
another solution could be to call  tiramisu::prepare_schedules_for_legality_checks() only when called from tiramisu::perform_full_dependency_analysis() 
so the part exposed to regular users would be more simple and those using the function object would need the difference.
so what do you think is the best approach ?

Reply:
Ok, let's stick with your 1st solution (i.e., let the user call it). I would rephrase the documentation as follows though (to make the requirements clearer):

Before invoking this method, the user must call tiramisu::prepare_schedules_for_legality_checks() and must define the buffer associated with each computation.